add best practice for secret expression (#4630)
harness/developer-hub: docs/platform/variables-and-expressions/harness-variables.md

  
  To invoke the method `concat` on the expression `<+pipeline.variables.var1>`, you must wrap `<+pipeline.variables.var1>` within `<+...>` and then invoke the method using `.concat()`.

> - While using an expression for secret functor provided by Harness `<+secrets.getValue("sec")>`, this expression should never be wrapped within quotation marks.

Reviewer: ```suggestion
-When using an expression for a Harness secret functor, '<+secrets.getValue("sec")>', it should not be wrapped within quotation marks.
```

>   Because this expression gets resolved to another internal functor of Harness `${ngSecretManager.obtain("sec")}`, which is resolved on delegate. Since, it's value is not a primitive like string, it should not be wrapped within quotation marks.

Reviewer: ```suggestion
  This expression gets resolved to another Harness internal functor,`${ngSecretManager.obtain("sec")}`, which is resolved on the delegate. Since its value is not a primitive type string, it should not be wrapped within quotation marks.
```

  
  To invoke the method `concat` on the expression `<+pipeline.variables.var1>`, you must wrap `<+pipeline.variables.var1>` within `<+...>` and then invoke the method using `.concat()`.

> -When using an expression for a Harness secret functor, '<+secrets.getValue("sec")>', it should not be wrapped within quotation marks.

Reviewer 2: I think you need to elaborate what a "Harness Secret Functor" is - is this a section in the Expressions reference? Is there a complete list? Or, if there is just one functor, let's elaborate on what it does.

Author: Harness secret functor is ` <+secrets.getValue("sec")>` only

> -When using an expression for a Harness secret functor, '<+secrets.getValue("sec")>', it should not be wrapped within quotation marks.

Reviewer: ```suggestion
- When using an expression for the Harness secret functor, `<+secrets.getValue("sec")>`, it should not be wrapped within quotation marks.
```

  
  This expression gets resolved to another Harness internal functor,`${ngSecretManager.obtain("sec")}`, which is resolved on the delegate. Since its value is not a primitive type string, it should not be wrapped within quotation marks.

  For example, consider the following expression for the value of a pipeline or stage variable:
  ```
  <+<+<+pipeline.variables.var1>=="secret1">?<+secrets.getValue("secret1")>:<+secrets.getValue("defaultSecret")>>
  ```
  This secret expression should not be wrapped within quotation marks.

> - If some expressions are not needed to be evaluated in pipeline yaml but are added as comments in the script in shell script step or run step etc., they will still be processed and evaluated, and it could lead to failure and unnecessary processing.

Reviewer: ```suggestion
- If expressions don't need to be evaluated in the pipeline YAML but are added as script comments in the Shell Script step, the Run step, or another step, they will still be processed and evaluated. This might cause failures and unnecessary processing.
```

> - If expressions don't need to be evaluated in the pipeline YAML but are added as script comments in the Shell Script step, the Run step, or another step, they will still be processed and evaluated. This might cause failures and unnecessary processing.

Reviewer: ```suggestion
- If expressions don't need to be evaluated in the pipeline YAML but are added as script comments in the Shell Script step, the Run step, or another step, they will still be processed and evaluated. This might cause failures and unnecessary processing.  Review and remove any unnecessary script comments from the pipeline YAML to streamline the evaluation process.
```

>   So, you should remove those commented code of having expressions if you don't want them to be evaluated.

Reviewer: ```suggestion
```
